src/lib.rs

```diff
@@ -44,17 +50,9 @@ pub fn lzma_decompress_with_options<R: io::BufRead, W: io::Write>(
     options: &decompress::Options,
 ) -> error::Result<()> {
     let params = decode::lzma::LzmaParams::read_header(input, options)?;
-    let mut decoder = decode::lzma::DecoderState::new(params.properties, params.unpacked_size);
-    let mut output = LzCircularBuffer::from_stream(
-        output,
-        params.dict_size as usize,
-        options.memlimit.unwrap_or(usize::MAX),
-    );
-
-    let mut rangecoder = decode::rangecoder::RangeDecoder::new(input)
-        .map_err(|e| error::Error::LzmaError(format!("LZMA stream too short: {}", e)))?;
-    decoder.process(&mut output, &mut rangecoder)?;
-    output.finish()?;
+    let mut decoder = decode::lzma::LzmaDecoder::new(params, options.memlimit)?;
+
+    decoder.decompress(input, output)?;
     Ok(())
```

Comment:
This can be simplified by having the function directly return the result from `decompress()`.